pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -238,7 +248,19 @@ func (rc *reconciler) mountAttachVolumes() {
 				}
 			}
 		} else if !volMounted || cache.IsRemountRequiredError(err) {
-			// Volume is not mounted, or is already mounted, but requires remounting
+			if !volMounted && rc.podStateProvider.ShouldPodContainersBeTerminating(volumeToMount.Pod.UID) {
+				// the volume is not yet mounted, and the pod containers should be terminating, so don't try to set up a new mount
+				// ShouldPodContainersBeTerminating: Intended for use by subsystem sync loops to avoid performing background setup after termination has been requested for a pod.
+				klog.V(4).InfoS(volumeToMount.GenerateMsgDetailed("Skipping operationExecutor.MountVolume because pod containers should be terminating", ""), "pod", klog.KObj(volumeToMount.Pod))
+				continue
+			}
```

Comment:
I'm less sure about this bit. While it seems silly to set up _new_ mounts for pods which have already been asked to terminate, and "avoid performing background setup after termination has been requested" certainly seems to match skipping here, I'm unsure if this will block syncPod and make pod termination take longer.

If a volume is not yet mounted, does that imply the containers depending on that volume are not yet running? If those containers are not yet running, and volume setup is in progress, and pod/container termination is requested, does the volume manager stop trying to set up the volumes or still block syncPod?

Reply:
I was thinking if this can cause a problem if a volume was added to DSOW and then pod got deleted but volume made it to ASOW because a previous mount partially timed out but we did not record the volume mount (or mounted in uncertain state), in which case we will never record volume was mounted. But this can happen without this change too - so we should be fine. 

> If a volume is not yet mounted, does that imply the containers depending on that volume are not yet running?

Yes - I think so. pod worker waits for all volumes to be mounted before starting the containers. And if the volume is not mounted, it should not block syncPod anymore.

Reply:
I also remember the logic [here](https://github.com/kubernetes/kubernetes/blob/aa0ea624890adace25b6f83dc121d339bff12b69/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L254). It will block volume from deleting from desired state if volume is not yet mounted to actual state once. The reason this check is added to avoid the situation that kubelet restarts, and mount information is not recorded in to actual state yet. 
With recent pod state information, I am wondering that logic should be updated too.
 causing mounts leaking problem

Reply:
I was thinking we can handle volume removing from desired state properly, it is the same effect of skipping mount when it is not necessary?

If a volume is added to desired state, but not yet in actual state. If pod is no longer needed, normally it should be removed from desired state, then reconciler will not trigger mount.

However, the logic is added to not removing it from desired state in some cases if volume is not yet in actual state. The reason is in case kubelet restarts, we might lose this volume information and cause mount leak. But on the other hand, I think it might cause the issue of pod stuck in terminating if volume cannot be successfully added to actual state forever. 

Is the change trying to help pod start or pod terminate successfully?

Reply:
> it is the same effect of skipping mount when it is not necessary

not quite... attempting to set up the mount here can require a successful get/read of an API object (like a PVC or secret or configmap) which may not exist or be accessible any more. We don't want to keep trying and failing to set up a new mount that isn't actually needed

Reply:
> > it is the same effect of skipping mount when it is not necessary
> 
> not quite... attempting to set up the mount here can require a successful get/read of an API object (like a PVC or secret or configmap) which may not exist or be accessible any more. We don't want to keep trying and failing to set up a new mount that isn't actually needed

My thought is that if volume can be removed from desired state in this case, the reconciler will not even need to try to set up the mount. The reason mount failed here is because volume is not removed from desired state due to the concern I mentioned above, so reconciler still try to mount it.

> 
> > I think it might cause the issue of pod stuck in terminating if volume cannot be successfully added to actual state forever.
> 
> I think failure to set up the mount here can also contribute to that... maybe that's the same thing you are saying?

So in current logic without this change, volume is added to desired state, but not yet in actual state. At this time, pod is deleting, since volume will not removed from desired state yet, so reconciler keeps mounting and failed. After pod is finally removed from pod manager, I think volume will be removed from desired state, the reconciler will stop mounting.

With this change, reconciler will skip unnecessary mounting. In the end, the same as above, pod should be removed from pod manager and volume will be removed from desired state.

So the unnecessary mounting is temporal, but it get/read API. With this fix, we get rid of this behavior, right?

Reply:
I'm not very familiar with all the retry/timeout layers that call here. If this returns without attempting a desired state mount because the pod containers should be terminating, will this function get re-called since dsw != asw, or will it not get re-called because the pod containers should be terminating?

Reply:
Volume reconciler just compare dsw and asw, as long as dsw != asw, it will trigger mount/unmount. desired_state_populator handles how to add/remove volume from the desired state. As long as volume is still in desired state, reconciler will try to mount no matter what pod state is.

Desired_state_populator has its own logic to decide whether to remove volume from it. It will check pod status there and also check whether the volume has be mounted or not (in actual state or not) https://github.com/kubernetes/kubernetes/blob/aa0ea624890adace25b6f83dc121d339bff12b69/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L245

Reply:
I think concern @jingxu97 has is - the following check:

```
if !volMounted && rc.podStateProvider.ShouldPodContainersBeTerminating(volumeToMount.Pod.UID) {
.... do not create mount
}

```

is not definitive. In case of kubelet's restarts - the volume could be mounted on the node, but ASW may not have that state *yet*, so mounting needs to continue so as ASW state can be rebuild correctly after kubelet restart(even if pod is terminating). if we skip mounting here, the volume *may* not make it into ASW and hence mount point may leak. There are fallbacks in volume reconstruction logic to possibly guard against that but reading the code in some cases they may not save us (and is race prone).

Reply:
From: https://github.com/kubernetes/kubernetes/pull/107739#issuecomment-1022842146
> background mount operation is still attempted and triggers the "not registered" error:
> 
> ```
> 21:48:40.292625 reconciler.go:247] "Starting operationExecutor.MountVolume for volume \"kube-api-access-jqksm\" (UniqueName: \"kubernetes.io/projected/5c414b9d-c754-40a8-9739-35e4feecde91-kube-api-access-jqksm\") pod \"deployment-8d545c96d-8nbtn\" (UID: \"5c414b9d-c754-40a8-9739-35e4feecde91\") " pod="apply-4355/deployment-8d545c96d-8nbtn"
> 21:48:40.328257 desired_state_of_world_populator.go:258] "Removing volume from desired state" pod="apply-4355/deployment-8d545c96d-8nbtn" podUID=5c414b9d-c754-40a8-9739-35e4feecde91 volumeName="kube-api-access-jqksm"
> 21:48:40.446359 watch_based_manager.go:297] "configmap.objectCache#Get not registered" namespace="apply-4355" name="kube-root-ca.crt"
> 21:48:40.446530 projected.go:293] Couldn't get configMap apply-4355/kube-root-ca.crt: object "apply-4355"/"kube-root-ca.crt" not registered
> ```
I think that a volume may just be removed from DSW between Line 200 (https://github.com/kubernetes/kubernetes/blob/d5263feb038825197ab426237b111086822366be/pkg/kubelet/volumemanager/reconciler/reconciler.go#L200) and Line 240 (https://github.com/kubernetes/kubernetes/blob/d5263feb038825197ab426237b111086822366be/pkg/kubelet/volumemanager/reconciler/reconciler.go#L240) when it was still existing before in` for _, volumeToMount := range rc.desiredStateOfWorld.GetVolumesToMount() `, so shall we check volume existing in dsw again here to prevent unneccessary mount? Since the volume mount is driven by dsw, if the volume is removed from dsw, the reconcile would unmout it in next loop immediately.